src/main/java/com/spotify/reaper/resources/RepairRunResource.java

```diff
@@ -68,6 +188,168 @@ public Response getRepairRunsForCluster(@PathParam("cluster_name") String cluste
+   * Splits a token range for given table into segments
+   * @return
+   * @throws ReaperException when fails to discover seeds for the cluster or fails to connect to
+   *   any of the nodes in the Cluster.
+   */
+  private List<RingRange> generateSegments(Cluster targetCluster, ColumnFamily existingTable)
+      throws ReaperException {
+    List<RingRange> segments = null;
+    SegmentGenerator sg = new SegmentGenerator(targetCluster.getPartitioner());
+    Set<String> seedHosts = targetCluster.getSeedHosts();
+    if (seedHosts.isEmpty()) {
+      String errMsg = String.format("didn't get any seed hosts for cluster \"%s\"",
+          existingTable.getClusterName());
+      LOG.error(errMsg);
+      throw new ReaperException(errMsg);
+    }
+    for (String host : seedHosts) {
+      try {
+        JmxProxy jmxProxy = jmxFactory.create(host);
+        List<BigInteger> tokens = jmxProxy.getTokens();
+        segments = sg.generateSegments(existingTable.getSegmentCount(), tokens);
+        jmxProxy.close();
+        break;
+      } catch (ReaperException e) {
+        LOG.warn("couldn't connect to host: {}, will try next one", host);
+      }
+    }
+    if (segments == null) {
+      String errMsg = String.format("failed to generate repair segments for cluster \"%s\"",
+          existingTable.getClusterName());
+      LOG.error(errMsg);
+      throw new ReaperException(errMsg);
+    }
+    return segments;
+  }
+
+  /**
+   * Instantiates a RepairRun and stores it in the storage backend.
+   * @return
+   * @throws ReaperException when fails to store the RepairRun.
+   */
+  private RepairRun storeNewRepairRun(Cluster cluster, ColumnFamily table, Optional<String> cause,
+      String owner) throws ReaperException {
+    RepairRun.Builder runBuilder = new RepairRun.Builder(cluster.getName(), table.getId(),
+        DateTime.now(), config.getRepairIntensity());
+    runBuilder.cause(cause.isPresent() ? cause.get() : "no cause specified");
+    runBuilder.owner(owner);
+    RepairRun newRepairRun = storage.addRepairRun(runBuilder);
+    if (newRepairRun == null) {
+      String errMsg = String.format("failed storing repair run for cluster \"%s/%s/%s",
+          cluster.getName(), table.getKeyspaceName(), table.getName());
+      LOG.error(errMsg);
+      throw new ReaperException(errMsg);
+    }
+    return newRepairRun;
+  }
+
+  /**
+   * Creates the repair runs linked to given RepairRun and stores them directly in the
+   * storage backend.
+   */
+  private void storeNewRepairSegments(List<RingRange> tokenSegments, RepairRun repairRun,
+      ColumnFamily table) {
+    List <RepairSegment.Builder> repairSegmentBuilders = Lists.newArrayList();
+    for (RingRange range : tokenSegments) {
+      RepairSegment.Builder repairSegment = new RepairSegment.Builder(repairRun.getId(), range,
+          table.getId());
+      repairSegmentBuilders.add(repairSegment);
+    }
+    // TODO(zvo): I don't like we can't figure out if this suceeds or not
```

Comment:
If the method call returns without throwing an exception, it means that the storage has processed and responded with success status.